sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/hcatalog/DatabaseProvider.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.sql.meta.provider.hcatalog;
+
+import com.alibaba.fastjson.JSONObject;
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.extensions.sql.BeamSqlTable;
+import org.apache.beam.sdk.extensions.sql.meta.Table;
+import org.apache.beam.sdk.extensions.sql.meta.provider.TableProvider;
+import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.hadoop.hive.metastore.api.NoSuchObjectException;
+
+/**
+ * Metastore has a structure of 'db.table'.
+ *
+ * <p>This provider represents the 'db' and contains the table look up logic.
+ */
+class DatabaseProvider implements TableProvider {
+
+  private String db;
+  private IMetaStoreClient metastore;
+  private Map<String, String> config;
+
+  DatabaseProvider(String db, IMetaStoreClient metastore, Map<String, String> config) {
+    this.db = db;
+    this.metastore = metastore;
+    this.config = config;
+  }
+
+  @Override
+  public String getTableType() {
+    return "hcatalog";
```

Comment:
It's not consistent that the provider named `Database` but the table type is `hcatalog`.

Reply:
The name of the provider represents the fact that it resolves the tables from a specific database. Table type is the kind of io/data that a table will represent, in this case we're reading `HCatRecords` from `HCatalog` using `HCatalogIO`

Reply:
I will see if I can come up with better naming

Reply:
This cannot be used with the shell yet, but I am planning to implement something like `REGISTER EXTERNAL SCHEMA hive TYPE hcatalog PROPERTIES { metastoreUris : ... }` similar to #6245

Reply:
At this moment, as we have the first table provider that will not be registered through auto service loader. We should start from this PR to separate interfaces between auto loaded table providers and manually registered table providers. These two category are mixed together right now. A bunch of providers are [loaded](https://github.com/apache/beam/blob/master/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/BeamCalciteSchemaFactory.java#L84) to both SQL shell and SqlTransform/BeamSqlEnv. This new provider implements TableProvider too.

It would be better to keep only one approach to register providers (so we don't maintain two code path), but in reality, SQL shell requires auto service loading approach to satisfy basic use case while not all providers can be registered through auto service because of difference of initialization parameters.

A practical approach will be one that makes all auto loaded providers use the same abstract class (It happens to be [InMemoryMetaTableProvider](https://github.com/apache/beam/blob/master/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/InMemoryMetaTableProvider.java) in current code base)  for auto loading only. And auto service loader will load all providers with annotation on that specific class.

Reply:
I do not think this belongs to this PR, the refactor to split the interfaces is going to be a larger scope. Also I do not think we have a clear understanding of what separation should look like.

For example, I believe that registration should not be part of the business model. Primary driver for class hierarchy should be the business model, not the way how we register things, i.e. if something provides the tables, it is a table provider no matter how it is registered, it should not be aware of the way it is registered, it's not its' concern. Registration of table providers is a configuration that happens in a separate place by a component who's concern is registration of table providers, for example. What if someone decides to use guice tomorrow to inject table providers?

I don't think that using `@AutoService` registration is a problem by itself. We have a specific part of our shell framework that loads specific table providers that are marked with `@AutoService`. I think it's a valid use case. Shell wants to auto-load some providers and doesn't want to auto-load others. The problem is that it's not configured in shell-specific-manner and basically anyone can use this annotation to load the classes. I think we do need to refactor it but I don't think this is something that we need to fix right away as I don't see actual issues being caused by this.

And I don't know what the solution should be. One way would be to use a registry approach similar to pipeline options, so that we only auto-load registries, not table providers themselves. This way shell needs to provide a registry, external implementations will also provide the registries for their table providers, etc. And if there is no registry to include a table provider, it's not auto-loaded. This makes configuration separate from the business model completely.

Or we could make users (people who build shell distribution or run it) explicitly configure the used table providers at run-time.

Or both.

And a separate problem is that we have a mixture of classes that provide tables, schemas, delegate to other providers, etc. E.g. I think we can move a bunch of table provider and Table logic into `BeamCalciteSchema` and this way we can split the schema provider from the actual data provider logic.

Reply:
I am not sure what business model means here which drives class hierarchy. AutoService loading itself is ok and and it's a valid use case for SQL shell.

The problem comes from: we are having multiple schema and more builtin providers while we still have auto service loading path to register table providers specifically to `beam` schema. It's causing race condition if users annotate their providers but also try to manually register those providers, which will win? At this moment users have to check implementation to figure it out if we don't have clear limitation or explanation on these approaches . The root cause is there is a loading order of providers  I can imagine one day if the order of initialization of schema changes by a PR, suddenly users find their new pipelines starting to have `table not found error`.  

This problem appeared after #7491 when we started to expose API for manually registering providers. If this PR does not fix it, at least this PR should be blocked until we have a fix.

If we have a package-private interface/abstract class for auto loaded table providers, the problem is more or less fixed.

Reply:
I thought `SqlTransform` uses auto service loading approach by default. I was wrong on it as `SqlTransform` does not use that. So my concern of potential race condition is not valid.